[quant][graphmode][refactor] Get QScheme from observer module

jerryzh168 opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack):
* #31408 [quant][graphmode] `insert_quant_dequant` pass support shared class types
* #31407 [quant][graphmode][refactor] Remove observers in the end
* #31406 [quant][graphmode][refactor] Call `getQSchemeAndQParamMap` later in `quantizeTensors`
* **#31293 [quant][graphmode][refactor] Get QScheme from observer module**

Summary:
Previously we check the number of elements in scale to determine if we are using per channel quantization,
but we should get qscheme information from observer module directly and we'll expose this information
to caller as well

Test Plan:
.

Reviewers:
mvz

Subscribers:

Tasks:

Tags:

Differential Revision: [D19146669](https://our.internmc.facebook.com/intern/diff/D19146669)

jerryzh168 committed:
[quant][graphmode][refactor] Exposing qscheme from observer module

Summary:
Previously we check the number of elements in scale to determine if we are using per channel quantization,
but we should get qscheme information from observer module directly and we'll expose this information
to caller as well

Test Plan:
.

Reviewers:
mvz

Subscribers:

Tasks:

Tags:

[ghstack-poisoned]

kostmo commented:
## CircleCI build failures summary
As of commit f6ecc93:
* **2/17** broken upstream at merge base 20a2e52 (see [grid view](https://dr.pytorch.org/master-timeline.html?count=1&sha1=20a2e526efb353e302f5b8778b63e45a5eade240&should_suppress_scheduled_builds=true&should_suppress_fully_successful_columns=true&max_columns_suppress_successful=35))
    * You may want to rebase on the `viable/strict` branch (see its [recency history](https://dr.pytorch.org/master-viable-commits.html)):
        * Since your merge base is older than `viable/strict`, run these commands:
          ```
          git fetch viable/strict
          git rebase viable/strict
          ```
* **15/17** failures introduced in this PR



## Detailed failure analysis

One may explore the probable reasons each build failed interactively [on the Dr. CI website](https://dr.pytorch.org/commit-details.html?sha1=f6ecc938d4d0b2c3ca8057cf372ef10d0f0eed22).

### 7 new failures recognized by patterns
The following build failures don't appear to be due to upstream breakage:
#### [![See CircleCI build](https://avatars0.githubusercontent.com/ml/7?s=12)](https://circleci.com/gh/pytorch/pytorch/3955834) pytorch_linux_xenial_py3_clang5_mobile_build (1/7)
**Step:** "Build" ([details](https://dr.pytorch.org/build-details.html?build_id=74479508))
<details>
<summary>
<code>Dec 15 01:48:28 /var/lib/jenkins/workspace/torch/csrc/jit/passes/quantization.cpp:771:42: error: cannot convert 'std::tuple<c10::QScheme, std::unordered_map<std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> >, c10::IValue, std::hash<std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> > >, std::equal_to<std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> > >, std::allocator<std::pair<const std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> >, c10::IValue> > > >' to 'int' in return</code>
</summary>

```
Dec 15 01:48:28                  from /var/lib/jenkins/workspace/torch/csrc/jit/passes/quantization.cpp:1: 
Dec 15 01:48:28 /var/lib/jenkins/workspace/c10/core/QScheme.h:15:12: note:   'c10::QScheme' 
Dec 15 01:48:28  enum class QScheme : uint8_t { 
Dec 15 01:48:28             ^ 
Dec 15 01:48:28 /var/lib/jenkins/workspace/c10/core/QScheme.h:15:12: note:   'c10::QScheme' 
Dec 15 01:48:28 /var/lib/jenkins/workspace/c10/core/QScheme.h:15:12: note:   'c10::QScheme' 
Dec 15 01:48:28 /var/lib/jenkins/workspace/torch/csrc/jit/passes/quantization.cpp:740:30: error: template argument 1 is invalid 
Dec 15 01:48:28  std::tuple<QScheme, QParamMap> InsertQuantDeQuantHelper::getQSchemeAndQParamMap( 
Dec 15 01:48:28                               ^ 
Dec 15 01:48:28 /var/lib/jenkins/workspace/torch/csrc/jit/passes/quantization.cpp: In member function 'int torch::jit::{anonymous}::InsertQuantDeQuantHelper::getQSchemeAndQParamMap(torch::jit::script::Module&, torch::jit::Value*)': 
ror: cannot convert 'std::tuple<c10::QScheme, std::unordered_map<std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> >, c10::IValue, std::hash<std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> > >, std::equal_to<std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> > >, std::allocator<std::pair<const std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> >, c10::IValue> > > >' to 'int' in return 
Dec 15 01:48:28    return std::make_tuple(qscheme, qparams); 
Dec 15 01:48:28                                           ^ 
Dec 15 01:48:28 caffe2/CMakeFiles/torch_cpu.dir/build.make:8536: recipe for target 'caffe2/CMakeFiles/torch_cpu.dir/__/torch/csrc/jit/passes/quantization.cpp.o' failed 
Dec 15 01:48:28 make[2]: *** [caffe2/CMakeFiles/torch_cpu.dir/__/torch/csrc/jit/passes/quantization.cpp.o] Error 1 
Dec 15 01:48:28 make[1]: *** [caffe2/CMakeFiles/torch_cpu.dir/all] Error 2 
Dec 15 01:48:28 CMakeFiles/Makefile2:882: recipe for target 'caffe2/CMakeFiles/torch_cpu.dir/all' failed 
Dec 15 01:48:28 Makefile:138: recipe for target 'all' failed 
Dec 15 01:48:28 make: *** [all] Error 2 
Dec 15 01:48:28 + sccache_epilogue 
Dec 15 01:48:28 + echo '=================== sccache compilation log ===================' 
```

</details>

#### [![See CircleCI build](https://avatars0.githubusercontent.com/ml/7?s=12)](https://circleci.com/gh/pytorch/pytorch/3955845) caffe2_onnx_py3_6_clang7_ubuntu16_04_build (2/7)
**Step:** "Build" ([details](https://dr.pytorch.org/build-details.html?build_id=74479779))
<details>
<summary>
<code>Dec 15 01:49:21 /var/lib/jenkins/workspace/torch/csrc/jit/passes/quantization.cpp:771:10: error: no viable conversion from returned value of type 'tuple<typename __decay_and_strip<QScheme &>::__type, typename __decay_and_strip<unordered_map<basic_string<char>, IValue, hash<string>, equal_to<basic_string<char> >, allocator<pair<const basic_string<char>, IValue> > > &>::__type>' (aka 'tuple<c10::QScheme, std::unordered_map<std::__cxx11::basic_string<char>, c10::IValue, std::hash<std::__cxx11::string>, std::equal_to<std::__cxx11::basic_string<char> >, std::allocator<std::pair<const std::__cxx11::basic_string<char>, c10::IValue> > > >') to function return type 'int'</code>
</summary>

```
Dec 15 01:49:21   if (tuple->elements().size() != num_outputs) { 
Dec 15 01:49:21       ~~~~~~~~~~~~~~~~~~~~~~~~ ^  ~~~~~~~~~~~ 
Dec 15 01:49:21 1 warning generated. 
Dec 15 01:49:21 [ 78%] Building CXX object caffe2/CMakeFiles/torch_cpu.dir/__/torch/csrc/jit/mobile/import.cpp.o 
Dec 15 01:49:21 /var/lib/jenkins/workspace/torch/csrc/jit/passes/quantization.cpp:624:14: error: use of undeclared identifier 'QScheme' 
Dec 15 01:49:21   std::tuple<QScheme, QParamMap> getQSchemeAndQParamMap(script::Module& module, Value* v); 
Dec 15 01:49:21              ^ 
Dec 15 01:49:21 /var/lib/jenkins/workspace/torch/csrc/jit/passes/quantization.cpp:740:12: error: use of undeclared identifier 'QScheme' 
Dec 15 01:49:21 std::tuple<QScheme, QParamMap> InsertQuantDeQuantHelper::getQSchemeAndQParamMap( 
Dec 15 01:49:21            ^ 
nd_strip<QScheme &>::__type, typename __decay_and_strip<unordered_map<basic_string<char>, IValue, hash<string>, equal_to<basic_string<char> >, allocator<pair<const basic_string<char>, IValue> > > &>::__type>' (aka 'tuple<c10::QScheme, std::unordered_map<std::__cxx11::basic_string<char>, c10::IValue, std::hash<std::__cxx11::string>, std::equal_to<std::__cxx11::basic_string<char> >, std::allocator<std::pair<const std::__cxx11::basic_string<char>, c10::IValue> > > >') to function return type 'int' 
Dec 15 01:49:21   return std::make_tuple(qscheme, qparams); 
Dec 15 01:49:21          ^~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~ 
Dec 15 01:49:21 3 errors generated. 
Dec 15 01:49:21 caffe2/CMakeFiles/torch_cpu.dir/build.make:20050: recipe for target 'caffe2/CMakeFiles/torch_cpu.dir/__/torch/csrc/jit/passes/quantization.cpp.o' failed 
Dec 15 01:49:21 make[2]: *** [caffe2/CMakeFiles/torch_cpu.dir/__/torch/csrc/jit/passes/quantization.cpp.o] Error 1 
Dec 15 01:49:21 make[2]: *** Waiting for unfinished jobs.... 
Dec 15 01:49:21 CMakeFiles/Makefile2:5548: recipe for target 'caffe2/CMakeFiles/torch_cpu.dir/all' failed 
Dec 15 01:49:21 make[1]: *** [caffe2/CMakeFiles/torch_cpu.dir/all] Error 2 
Dec 15 01:49:21 Makefile:138: recipe for target 'all' failed 
Dec 15 01:49:21 make: *** [all] Error 2 
```

</details>

#### [![See CircleCI build](https://avatars0.githubusercontent.com/ml/7?s=12)](https://circleci.com/gh/pytorch/pytorch/3955847) caffe2_onnx_py2_gcc5_ubuntu16_04_build (3/7)
**Step:** "Build" ([details](https://dr.pytorch.org/build-details.html?build_id=74479888))
<details>
<summary>
<code>Dec 15 01:49:38 /var/lib/jenkins/workspace/torch/csrc/jit/passes/quantization.cpp:771:42: error: cannot convert 'std::tuple<c10::QScheme, std::unordered_map<std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> >, c10::IValue, std::hash<std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> > >, std::equal_to<std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> > >, std::allocator<std::pair<const std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> >, c10::IValue> > > >' to 'int' in return</code>
</summary>

```
Dec 15 01:49:38                  from /var/lib/jenkins/workspace/torch/csrc/jit/passes/quantization.cpp:1: 
Dec 15 01:49:38 /var/lib/jenkins/workspace/c10/core/QScheme.h:15:12: note:   'c10::QScheme' 
Dec 15 01:49:38  enum class QScheme : uint8_t { 
Dec 15 01:49:38             ^ 
Dec 15 01:49:38 /var/lib/jenkins/workspace/c10/core/QScheme.h:15:12: note:   'c10::QScheme' 
Dec 15 01:49:38 /var/lib/jenkins/workspace/c10/core/QScheme.h:15:12: note:   'c10::QScheme' 
Dec 15 01:49:38 /var/lib/jenkins/workspace/torch/csrc/jit/passes/quantization.cpp:740:30: error: template argument 1 is invalid 
Dec 15 01:49:38  std::tuple<QScheme, QParamMap> InsertQuantDeQuantHelper::getQSchemeAndQParamMap( 
Dec 15 01:49:38                               ^ 
Dec 15 01:49:38 /var/lib/jenkins/workspace/torch/csrc/jit/passes/quantization.cpp: In member function 'int torch::jit::{anonymous}::InsertQuantDeQuantHelper::getQSchemeAndQParamMap(torch::jit::script::Module&, torch::jit::Value*)': 
ror: cannot convert 'std::tuple<c10::QScheme, std::unordered_map<std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> >, c10::IValue, std::hash<std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> > >, std::equal_to<std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> > >, std::allocator<std::pair<const std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> >, c10::IValue> > > >' to 'int' in return 
Dec 15 01:49:38    return std::make_tuple(qscheme, qparams); 
Dec 15 01:49:38                                           ^ 
Dec 15 01:49:38 caffe2/CMakeFiles/torch_cpu.dir/build.make:20050: recipe for target 'caffe2/CMakeFiles/torch_cpu.dir/__/torch/csrc/jit/passes/quantization.cpp.o' failed 
Dec 15 01:49:38 make[2]: *** [caffe2/CMakeFiles/torch_cpu.dir/__/torch/csrc/jit/passes/quantization.cpp.o] Error 1 
Dec 15 01:49:38 make[2]: *** Waiting for unfinished jobs.... 
Dec 15 01:49:38 make[1]: *** [caffe2/CMakeFiles/torch_cpu.dir/all] Error 2 
Dec 15 01:49:38 CMakeFiles/Makefile2:5422: recipe for target 'caffe2/CMakeFiles/torch_cpu.dir/all' failed 
Dec 15 01:49:38 make: *** [all] Error 2 
Dec 15 01:49:38 Makefile:138: recipe for target 'all' failed 
Dec 15 01:49:38 Traceback (most recent call last): 
```

</details>

#### [![See CircleCI build](https://avatars0.githubusercontent.com/ml/7?s=12)](https://circleci.com/gh/pytorch/pytorch/3955854) binary_linux_manywheel_2_7mu_cpu_devtoolset7_build (4/7)
**Step:** "Build" ([details](https://dr.pytorch.org/build-details.html?build_id=74481368))
<details>
<summary>
<code>Dec 15 01:54:20 /pytorch/torch/csrc/jit/passes/quantization.cpp:771:42: error: cannot convert ‘std::tuple<c10::QScheme, std::unordered_map<std::basic_string<char, std::char_traits<char>, std::allocator<char> >, c10::IValue, std::hash<std::basic_string<char, std::char_traits<char>, std::allocator<char> > >, std::equal_to<std::basic_string<char, std::char_traits<char>, std::allocator<char> > >, std::allocator<std::pair<const std::basic_string<char, std::char_traits<char>, std::allocator<char> >, c10::IValue> > > >’ to ‘int’ in return</code>
</summary>

```
Dec 15 01:54:20                  from /pytorch/torch/csrc/jit/passes/quantization.cpp:1: 
Dec 15 01:54:20 /pytorch/c10/core/QScheme.h:15:12: note:   ‘c10::QScheme’ 
Dec 15 01:54:20  enum class QScheme : uint8_t { 
Dec 15 01:54:20             ^~~~~~~ 
Dec 15 01:54:20 /pytorch/c10/core/QScheme.h:15:12: note:   ‘c10::QScheme’ 
Dec 15 01:54:20 /pytorch/c10/core/QScheme.h:15:12: note:   ‘c10::QScheme’ 
Dec 15 01:54:20 /pytorch/torch/csrc/jit/passes/quantization.cpp:740:30: error: template argument 1 is invalid 
Dec 15 01:54:20  std::tuple<QScheme, QParamMap> InsertQuantDeQuantHelper::getQSchemeAndQParamMap( 
Dec 15 01:54:20                               ^ 
Dec 15 01:54:20 /pytorch/torch/csrc/jit/passes/quantization.cpp: In member function ‘int torch::jit::{anonymous}::InsertQuantDeQuantHelper::getQSchemeAndQParamMap(torch::jit::script::Module&, torch::jit::Value*)’: 
t/passes/quantization.cpp:771:42: error: cannot convert ‘std::tuple<c10::QScheme, std::unordered_map<std::basic_string<char, std::char_traits<char>, std::allocator<char> >, c10::IValue, std::hash<std::basic_string<char, std::char_traits<char>, std::allocator<char> > >, std::equal_to<std::basic_string<char, std::char_traits<char>, std::allocator<char> > >, std::allocator<std::pair<const std::basic_string<char, std::char_traits<char>, std::allocator<char> >, c10::IValue> > > >’ to ‘int’ in return 
Dec 15 01:54:20    return std::make_tuple(qscheme, qparams); 
Dec 15 01:54:20                                           ^ 
Dec 15 01:54:22 [ 79%] Building CXX object caffe2/CMakeFiles/torch_cpu.dir/__/torch/csrc/jit/script/compiler.cpp.o 
Dec 15 01:54:23 [ 79%] Building CXX object caffe2/CMakeFiles/torch_cpu.dir/__/torch/csrc/jit/testing/file_check.cpp.o 
Dec 15 01:54:23 [ 79%] Building CXX object caffe2/CMakeFiles/torch_cpu.dir/__/torch/csrc/jit/script/convert_to_ssa.cpp.o 
Dec 15 01:54:23 [ 79%] Building CXX object caffe2/CMakeFiles/torch_cpu.dir/__/torch/csrc/jit/script/exit_transforms.cpp.o 
Dec 15 01:54:23 [ 79%] Building CXX object caffe2/CMakeFiles/torch_cpu.dir/__/torch/csrc/jit/script/inline_loop_condition.cpp.o 
Dec 15 01:54:24 gmake[2]: *** [caffe2/CMakeFiles/torch_cpu.dir/__/torch/csrc/jit/passes/quantization.cpp.o] Error 1 
Dec 15 01:54:24 gmake[2]: *** Waiting for unfinished jobs.... 
Dec 15 01:54:42 gmake[1]: *** [caffe2/CMakeFiles/torch_cpu.dir/all] Error 2 
```

</details>

#### [![See CircleCI build](https://avatars0.githubusercontent.com/ml/7?s=12)](https://circleci.com/gh/pytorch/pytorch/3955848) binary_linux_libtorch_2_7m_cpu_devtoolset7_shared-with-deps_build (5/7)
**Step:** "Build" ([details](https://dr.pytorch.org/build-details.html?build_id=74482047))
<details>
<summary>
<code>Dec 15 01:55:12 /pytorch/torch/csrc/jit/passes/quantization.cpp:771:42: error: cannot convert ‘std::tuple<c10::QScheme, std::unordered_map<std::basic_string<char, std::char_traits<char>, std::allocator<char> >, c10::IValue, std::hash<std::basic_string<char, std::char_traits<char>, std::allocator<char> > >, std::equal_to<std::basic_string<char, std::char_traits<char>, std::allocator<char> > >, std::allocator<std::pair<const std::basic_string<char, std::char_traits<char>, std::allocator<char> >, c10::IValue> > > >’ to ‘int’ in return</code>
</summary>

```
Dec 15 01:55:12                  from /pytorch/torch/csrc/jit/passes/quantization.cpp:1: 
Dec 15 01:55:12 /pytorch/c10/core/QScheme.h:15:12: note:   ‘c10::QScheme’ 
Dec 15 01:55:12  enum class QScheme : uint8_t { 
Dec 15 01:55:12             ^~~~~~~ 
Dec 15 01:55:12 /pytorch/c10/core/QScheme.h:15:12: note:   ‘c10::QScheme’ 
Dec 15 01:55:12 /pytorch/c10/core/QScheme.h:15:12: note:   ‘c10::QScheme’ 
Dec 15 01:55:12 /pytorch/torch/csrc/jit/passes/quantization.cpp:740:30: error: template argument 1 is invalid 
Dec 15 01:55:12  std::tuple<QScheme, QParamMap> InsertQuantDeQuantHelper::getQSchemeAndQParamMap( 
Dec 15 01:55:12                               ^ 
Dec 15 01:55:12 /pytorch/torch/csrc/jit/passes/quantization.cpp: In member function ‘int torch::jit::{anonymous}::InsertQuantDeQuantHelper::getQSchemeAndQParamMap(torch::jit::script::Module&, torch::jit::Value*)’: 
t/passes/quantization.cpp:771:42: error: cannot convert ‘std::tuple<c10::QScheme, std::unordered_map<std::basic_string<char, std::char_traits<char>, std::allocator<char> >, c10::IValue, std::hash<std::basic_string<char, std::char_traits<char>, std::allocator<char> > >, std::equal_to<std::basic_string<char, std::char_traits<char>, std::allocator<char> > >, std::allocator<std::pair<const std::basic_string<char, std::char_traits<char>, std::allocator<char> >, c10::IValue> > > >’ to ‘int’ in return 
Dec 15 01:55:12    return std::make_tuple(qscheme, qparams); 
Dec 15 01:55:12                                           ^ 
Dec 15 01:55:15 [ 80%] Building CXX object caffe2/CMakeFiles/torch_cpu.dir/__/torch/csrc/jit/script/compiler.cpp.o 
Dec 15 01:55:15 [ 80%] Building CXX object caffe2/CMakeFiles/torch_cpu.dir/__/torch/csrc/jit/testing/file_check.cpp.o 
Dec 15 01:55:15 [ 80%] Building CXX object caffe2/CMakeFiles/torch_cpu.dir/__/torch/csrc/jit/script/convert_to_ssa.cpp.o 
Dec 15 01:55:15 [ 80%] Building CXX object caffe2/CMakeFiles/torch_cpu.dir/__/torch/csrc/jit/script/exit_transforms.cpp.o 
Dec 15 01:55:16 gmake[2]: *** [caffe2/CMakeFiles/torch_cpu.dir/__/torch/csrc/jit/passes/quantization.cpp.o] Error 1 
Dec 15 01:55:16 gmake[2]: *** Waiting for unfinished jobs.... 
Dec 15 01:55:36 gmake[1]: *** [caffe2/CMakeFiles/torch_cpu.dir/all] Error 2 
Dec 15 01:55:36 gmake: *** [all] Error 2 
```

</details>

#### [![See CircleCI build](https://avatars0.githubusercontent.com/ml/7?s=12)](https://circleci.com/gh/pytorch/pytorch/3955840) pytorch_windows_build (6/7)
**Step:** "Build" ([details](https://dr.pytorch.org/build-details.html?build_id=74483829))
<details>
<summary>
<code>FAILED: caffe2/CMakeFiles/torch_cpu.dir/__/torch/csrc/jit/passes/quantization.cpp.obj</code>
</summary>

```

EFINITION -DHAVE_AVX2_CPU_DEFINITION /MD /O2 /Ob2  /w /EHa /MP /bigobj   -DCUDA_HAS_FP16=1 -DUSE_GCC_GET_CPUID -DUSE_AVX -DUSE_AVX2 -DTH_HAVE_THREAD /Z7 /EHa /DNOMINMAX /wd4267 /wd4251 /wd4522 /wd4838 /wd4305 /wd4244 /wd4190 /wd4101 /wd4996 /wd4275 /bigobj -O2 -DCAFFE2_BUILD_MAIN_LIB -DONNX_BUILD_MAIN_LIB -std:c++14 /showIncludes /Focaffe2\CMakeFiles\torch_cpu.dir\__\torch\csrc\jit\script\jit_exception.cpp.obj /Fdcaffe2\CMakeFiles\torch_cpu.dir\ /FS -c ..\torch\csrc\jit\script\jit_exception.cpp 
Microsoft (R) C/C++ Optimizing Compiler Version 19.23.28107 for x64
Copyright (C) Microsoft Corporation.  All rights reserved.

HAVE_AVX_CPU_DEFINITION -DHAVE_AVX2_CPU_DEFINITION /MD /O2 /Ob2  /w /EHa /MP /bigobj   -DCUDA_HAS_FP16=1 -DUSE_GCC_GET_CPUID -DUSE_AVX -DUSE_AVX2 -DTH_HAVE_THREAD /Z7 /EHa /DNOMINMAX /wd4267 /wd4251 /wd4522 /wd4838 /wd4305 /wd4244 /wd4190 /wd4101 /wd4996 /wd4275 /bigobj -O2 -DCAFFE2_BUILD_MAIN_LIB -DONNX_BUILD_MAIN_LIB -std:c++14 /showIncludes /Focaffe2\CMakeFiles\torch_cpu.dir\__\torch\csrc\jit\script\parser.cpp.obj /Fdcaffe2\CMakeFiles\torch_cpu.dir\ /FS -c ..\torch\csrc\jit\script\parser.cpp 
Microsoft (R) C/C++ Optimizing Compiler Version 19.23.28107 for x64
Copyright (C) Microsoft Corporation.  All rights reserved.

_DEFINITION -DHAVE_AVX2_CPU_DEFINITION /MD /O2 /Ob2  /w /EHa /MP /bigobj   -DCUDA_HAS_FP16=1 -DUSE_GCC_GET_CPUID -DUSE_AVX -DUSE_AVX2 -DTH_HAVE_THREAD /Z7 /EHa /DNOMINMAX /wd4267 /wd4251 /wd4522 /wd4838 /wd4305 /wd4244 /wd4190 /wd4101 /wd4996 /wd4275 /bigobj -O2 -DCAFFE2_BUILD_MAIN_LIB -DONNX_BUILD_MAIN_LIB -std:c++14 /showIncludes /Focaffe2\CMakeFiles\torch_cpu.dir\__\torch\csrc\jit\passes\quantization.cpp.obj /Fdcaffe2\CMakeFiles\torch_cpu.dir\ /FS -c ..\torch\csrc\jit\passes\quantization.cpp 
FAILED: caffe2/CMakeFiles/torch_cpu.dir/__/torch/csrc/jit/passes/quantization.cpp.obj  
_DEFINITION -DHAVE_AVX2_CPU_DEFINITION /MD /O2 /Ob2  /w /EHa /MP /bigobj   -DCUDA_HAS_FP16=1 -DUSE_GCC_GET_CPUID -DUSE_AVX -DUSE_AVX2 -DTH_HAVE_THREAD /Z7 /EHa /DNOMINMAX /wd4267 /wd4251 /wd4522 /wd4838 /wd4305 /wd4244 /wd4190 /wd4101 /wd4996 /wd4275 /bigobj -O2 -DCAFFE2_BUILD_MAIN_LIB -DONNX_BUILD_MAIN_LIB -std:c++14 /showIncludes /Focaffe2\CMakeFiles\torch_cpu.dir\__\torch\csrc\jit\passes\quantization.cpp.obj /Fdcaffe2\CMakeFiles\torch_cpu.dir\ /FS -c ..\torch\csrc\jit\passes\quantization.cpp 
..\torch\csrc\jit\passes\quantization.cpp(624): error C2065: 'QScheme': undeclared identifier
..\torch\csrc\jit\passes\quantization.cpp(624): error C3544: '_Types': parameter pack expects a type template argument
..\torch\csrc\jit\passes\quantization.cpp(661): error C2027: use of undefined type 'std::tuple'
C:\Program Files (x86)\Microsoft Visual Studio\2019\Community\VC\Tools\MSVC\14.23.28105\include\tuple(197): note: see declaration of 'std::tuple'
..\torch\csrc\jit\passes\quantization.cpp(662): error C2672: 'std::get': no matching overloaded function found
..\torch\csrc\jit\passes\quantization.cpp(662): error C2784: 'tuple_element<_Idx,std::pair<_Ty1,_Ty2>>::type &std::get(std::pair<_Ty1,_Ty2> &) noexcept': could not deduce template argument for 'std::pair<_Ty1,_Ty2> &' from 'std::tuple'
C:\Program Files (x86)\Microsoft Visual Studio\2019\Community\VC\Tools\MSVC\14.23.28105\include\utility(453): note: see declaration of 'std::get'
..\torch\csrc\jit\passes\quantization.cpp(663): error C3536: 'qparam_map': cannot be used before it is initialized
ue>>>,std::_Uhash_compare<_Kty,_Hasher,_Keyeq>,_Alloc,false>>::insert(std::pair<const _Kty,std::unordered_map<std::string,c10::IValue,std::hash<std::basic_string<char,std::char_traits<char>,std::allocator<char>>>,std::equal_to<std::basic_string<char,std::char_traits<char>,std::allocator<char>>>,std::allocator<std::pair<const std::basic_string<char,std::char_traits<char>,std::allocator<char>>,c10::IValue>>>> &&)': cannot convert argument 1 from 'initializer list' to 'std::pair<const _Kty,_Ty> &&'
```

</details>

#### [![See CircleCI build](https://avatars0.githubusercontent.com/ml/7?s=12)](https://circleci.com/gh/pytorch/pytorch/3955849) pytorch_linux_xenial_py3_clang5_android_ndk_r19c_x86_32_build (7/7)
**Step:** "Build" ([details](https://dr.pytorch.org/build-details.html?build_id=74485335))
<details>
<summary>
<code>Dec 15 02:08:58 FAILED: caffe2/CMakeFiles/torch_cpu.dir/__/torch/csrc/jit/passes/quantization.cpp.o</code>
</summary>

```
Dec 15 02:08:31 [533/581] Building CXX object caffe2/CMakeFiles/torch_cpu.dir/__/torch/csrc/jit/passes/requires_grad_analysis.cpp.o 
Dec 15 02:08:36 [534/581] Building CXX object caffe2/CMakeFiles/torch_cpu.dir/__/torch/csrc/jit/passes/specialize_autogradzero.cpp.o 
Dec 15 02:08:38 [535/581] Building CXX object caffe2/CMakeFiles/torch_cpu.dir/__/torch/csrc/jit/passes/subgraph_rewrite.cpp.o 
Dec 15 02:08:44 [536/581] Building CXX object caffe2/CMakeFiles/torch_cpu.dir/__/torch/csrc/jit/passes/utils/subgraph_utils.cpp.o 
Dec 15 02:08:46 [537/581] Building CXX object caffe2/CMakeFiles/torch_cpu.dir/__/torch/csrc/jit/passes/shape_analysis.cpp.o 
Dec 15 02:08:49 [538/581] Building CXX object caffe2/CMakeFiles/torch_cpu.dir/__/torch/csrc/jit/passes/utils/memory_dag.cpp.o 
Dec 15 02:08:55 [539/581] Building CXX object caffe2/CMakeFiles/torch_cpu.dir/__/torch/csrc/jit/passes/python_print.cpp.o 
Dec 15 02:08:55 [540/581] Building CXX object caffe2/CMakeFiles/torch_cpu.dir/__/torch/csrc/jit/passes/utils/check_alias_annotation.cpp.o 
Dec 15 02:08:56 [541/581] Building CXX object caffe2/CMakeFiles/torch_cpu.dir/__/torch/csrc/jit/print_handler.cpp.o 
Dec 15 02:08:58 [542/581] Building CXX object caffe2/CMakeFiles/torch_cpu.dir/__/torch/csrc/jit/passes/quantization.cpp.o 
Dec 15 02:08:58 FAILED: caffe2/CMakeFiles/torch_cpu.dir/__/torch/csrc/jit/passes/quantization.cpp.o  
rameter -Wno-missing-field-initializers -Wno-write-strings -Wno-unknown-pragmas -Wno-missing-braces -Wno-maybe-uninitialized -fvisibility=hidden -O2 -DCAFFE2_BUILD_MAIN_LIB -std=gnu++14 -MD -MT caffe2/CMakeFiles/torch_cpu.dir/__/torch/csrc/jit/passes/quantization.cpp.o -MF caffe2/CMakeFiles/torch_cpu.dir/__/torch/csrc/jit/passes/quantization.cpp.o.d -o caffe2/CMakeFiles/torch_cpu.dir/__/torch/csrc/jit/passes/quantization.cpp.o -c /var/lib/jenkins/workspace/torch/csrc/jit/passes/quantization.cpp 
Dec 15 02:08:58 /var/lib/jenkins/workspace/torch/csrc/jit/passes/quantization.cpp:624:14: error: use of undeclared identifier 'QScheme' 
Dec 15 02:08:58   std::tuple<QScheme, QParamMap> getQSchemeAndQParamMap(script::Module& module, Value* v); 
Dec 15 02:08:58              ^ 
Dec 15 02:08:58 /var/lib/jenkins/workspace/torch/csrc/jit/passes/quantization.cpp:740:12: error: use of undeclared identifier 'QScheme' 
Dec 15 02:08:58 std::tuple<QScheme, QParamMap> InsertQuantDeQuantHelper::getQSchemeAndQParamMap( 
Dec 15 02:08:58            ^ 
ndk1::char_traits<char>, std::__ndk1::allocator<char> >, c10::IValue, std::__ndk1::hash<std::__ndk1::basic_string<char, std::__ndk1::char_traits<char>, std::__ndk1::allocator<char> > >, std::__ndk1::equal_to<std::__ndk1::basic_string<char, std::__ndk1::char_traits<char>, std::__ndk1::allocator<char> > >, std::__ndk1::allocator<std::__ndk1::pair<const std::__ndk1::basic_string<char, std::__ndk1::char_traits<char>, std::__ndk1::allocator<char> >, c10::IValue> > > >') to function return type 'int' 
Dec 15 02:08:58   return std::make_tuple(qscheme, qparams); 
Dec 15 02:08:58          ^~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~ 
```

</details>

### 2 upstream failures recognized by patterns:
These builds matched patterns, but were probably caused by upstream breakages:
* [pytorch_linux_xenial_cuda9_cudnn7_py3_build](https://dr.pytorch.org/build-details.html?build_id=74482268)
* [pytorch_libtorch_linux_xenial_cuda9_cudnn7_py3_build](https://dr.pytorch.org/build-details.html?build_id=74483327)

### 8 failures *not* recognized by patterns:
| Job | Step | Status |
| --- | --- | --- |
| ![CircleCI](https://avatars0.githubusercontent.com/ml/7?s=12) <sup>pytorch_linux_xenial_py3_clang5_asan_build</sup> | <sup>Build</sup> | New in PR |
| ![CircleCI](https://avatars0.githubusercontent.com/ml/7?s=12) <sup>pytorch_linux_xenial_py3_6_gcc5_4_build</sup> | <sup>Build</sup> | New in PR |
| ![CircleCI](https://avatars0.githubusercontent.com/ml/7?s=12) <sup>pytorch_linux_xenial_py2_7_9_build</sup> | <sup>Build</sup> | New in PR |
| ![CircleCI](https://avatars0.githubusercontent.com/ml/7?s=12) <sup>pytorch_xla_linux_xenial_py3_6_clang7_build</sup> | <sup>Build</sup> | New in PR |
| ![CircleCI](https://avatars0.githubusercontent.com/ml/7?s=12) <sup>pytorch_macos_10_13_py3_build</sup> | <sup>Build</sup> | New in PR |
| ![CircleCI](https://avatars0.githubusercontent.com/ml/7?s=12) <sup>pytorch_macos_10_13_cuda10_0_cudnn7_py3_build</sup> | <sup>Build</sup> | New in PR |
| ![CircleCI](https://avatars0.githubusercontent.com/ml/7?s=12) <sup>binary_linux_libtorch_2_7m_cpu_gcc5_4_cxx11-abi_shared-with-deps_build</sup> | <sup>Build</sup> | New in PR |
| ![CircleCI](https://avatars0.githubusercontent.com/ml/7?s=12) <sup>binary_macos_libtorch_2_7_cpu_build</sup> | <sup>Build</sup> | New in PR |


---
This comment was automatically generated by [Dr. CI](https://github.com/kostmo/circleci-failure-tracker/tree/master/docs/from-pull-request-comment).
Follow [this link to opt-out](https://dr.pytorch.org/admin/comments-opt-out.html) of these comments for your Pull Requests.


Please report bugs/suggestions on the [GitHub issue tracker](https://github.com/kostmo/circleci-failure-tracker/issues).

jerryzh168 committed:
Update on "[quant][graphmode][refactor] Exposing qscheme from observer module"


Summary:
Previously we check the number of elements in scale to determine if we are using per channel quantization,
but we should get qscheme information from observer module directly and we'll expose this information
to caller as well

Test Plan:
.

Reviewers:
mvz

Subscribers:

Tasks:

Tags:

[ghstack-poisoned]

jerryzh168 committed:
Update on "[quant][graphmode][refactor] Get QScheme from observer module"


Summary:
Previously we check the number of elements in scale to determine if we are using per channel quantization,
but we should get qscheme information from observer module directly and we'll expose this information
to caller as well

Test Plan:
.

Reviewers:
mvz

Subscribers:

Tasks:

Tags:

[ghstack-poisoned]

jerryzh168 committed:
Update on "[quant][graphmode][refactor] Get QScheme from observer module"


Summary:
Previously we check the number of elements in scale to determine if we are using per channel quantization,
but we should get qscheme information from observer module directly and we'll expose this information
to caller as well

Test Plan:
.

Reviewers:
mvz

Subscribers:

Tasks:

Tags:

[ghstack-poisoned]

ZolotukhinM commented on torch/csrc/jit/passes/quantization.cpp:
```diff
@@ -735,13 +737,13 @@ void checkGetQParamsResult(const IValue& qparams) {
 }
 
-QParamMap InsertQuantDeQuantHelper::getQParamMap(
+std::tuple<QScheme, QParamMap> InsertQuantDeQuantHelper::getQSchemeAndQParamMap(
```
It doesn't look like we're ever using the first element of the tuple. Do we really need it?

jerryzh168 replied:
It's used in next PR :)

ZolotukhinM reviewed (approved):
Looks good modulo one question inline.

jerryzh168 committed:
Update on "[quant][graphmode][refactor] Get QScheme from observer module"


Summary:
Previously we check the number of elements in scale to determine if we are using per channel quantization,
but we should get qscheme information from observer module directly and we'll expose this information
to caller as well

Test Plan:
.

Reviewers:
mvz

Subscribers:

Tasks:

Tags:

Differential Revision: [D19146669](https://our.internmc.facebook.com/intern/diff/D19146669)

[ghstack-poisoned]

jerryzh168 committed:
Update on "[quant][graphmode][refactor] Get QScheme from observer module"


Summary:
Previously we check the number of elements in scale to determine if we are using per channel quantization,
but we should get qscheme information from observer module directly and we'll expose this information
to caller as well

Test Plan:
.

Reviewers:
mvz

Subscribers:

Tasks:

Tags:

Differential Revision: [D19146669](https://our.internmc.facebook.com/intern/diff/D19146669)

[ghstack-poisoned]

jerryzh168 committed:
Update on "[quant][graphmode][refactor] Get QScheme from observer module"


Summary:
Previously we check the number of elements in scale to determine if we are using per channel quantization,
but we should get qscheme information from observer module directly and we'll expose this information
to caller as well

Test Plan:
.

Reviewers:
mvz

Subscribers:

Tasks:

Tags:

Differential Revision: [D19146669](https://our.internmc.facebook.com/intern/diff/D19146669)

[ghstack-poisoned]

jerryzh168 committed:
Update on "[quant][graphmode][refactor] Get QScheme from observer module"


Summary:
Previously we check the number of elements in scale to determine if we are using per channel quantization,
but we should get qscheme information from observer module directly and we'll expose this information
to caller as well

Test Plan:
.

Reviewers:
mvz

Subscribers:

Tasks:

Tags:

Differential Revision: [D19146669](https://our.internmc.facebook.com/intern/diff/D19146669)

[ghstack-poisoned]

jerryzh168 committed:
Update on "[quant][graphmode][refactor] Get QScheme from observer module"


Summary:
Previously we check the number of elements in scale to determine if we are using per channel quantization,
but we should get qscheme information from observer module directly and we'll expose this information
to caller as well

Test Plan:
.

Reviewers:
mvz

Subscribers:

Tasks:

Tags:

Differential Revision: [D19146669](https://our.internmc.facebook.com/intern/diff/D19146669)

[ghstack-poisoned]

jerryzh168 committed:
Update on "[quant][graphmode][refactor] Get QScheme from observer module"


Summary:
Previously we check the number of elements in scale to determine if we are using per channel quantization,
but we should get qscheme information from observer module directly and we'll expose this information
to caller as well

Test Plan:
.

Reviewers:
mvz

Subscribers:

Tasks:

Tags:

Differential Revision: [D19146669](https://our.internmc.facebook.com/intern/diff/D19146669)

[ghstack-poisoned]

jerryzh168 committed:
Update on "[quant][graphmode][refactor] Get QScheme from observer module"


Summary:
Previously we check the number of elements in scale to determine if we are using per channel quantization,
but we should get qscheme information from observer module directly and we'll expose this information
to caller as well

Test Plan:
.

Reviewers:
mvz

Subscribers:

Tasks:

Tags:

Differential Revision: [D19146669](https://our.internmc.facebook.com/intern/diff/D19146669)

[ghstack-poisoned]

ZolotukhinM commented on torch/csrc/jit/passes/quantization.cpp:
```diff
@@ -623,7 +625,7 @@ class InsertQuantDeQuantHelper {
   // by searching for observer module of the value and extract the
   // quantization parameters from the observer module
-  QParamMap getQParamMap(script::Module& module, Value* v);
+  std::tuple<c10::QScheme, QParamMap> getQSchemeAndQParamMap(script::Module& module, Value* v);
```
This line is too long, please clang-format.

jerryzh168 committed:
Update on "[quant][graphmode][refactor] Get QScheme from observer module"


Summary:
Previously we check the number of elements in scale to determine if we are using per channel quantization,
but we should get qscheme information from observer module directly and we'll expose this information
to caller as well

Test Plan:
.

Reviewers:
mvz

Subscribers:

Tasks:

Tags:

Differential Revision: [D19146669](https://our.internmc.facebook.com/intern/diff/D19146669)

[ghstack-poisoned]

jerryzh168 committed:
Update on "[quant][graphmode][refactor] Get QScheme from observer module"


Summary:
Previously we check the number of elements in scale to determine if we are using per channel quantization,
but we should get qscheme information from observer module directly and we'll expose this information
to caller as well

Test Plan:
.

Reviewers:
mvz

Subscribers:

Tasks:

Tags:

Differential Revision: [D19146669](https://our.internmc.facebook.com/intern/diff/D19146669)

[ghstack-poisoned]

jerryzh168 committed:
Update on "[quant][graphmode][refactor] Get QScheme from observer module"


Summary:
Previously we check the number of elements in scale to determine if we are using per channel quantization,
but we should get qscheme information from observer module directly and we'll expose this information
to caller as well

Test Plan:
.

Reviewers:
mvz

Subscribers:

Tasks:

Tags:

Differential Revision: [D19146669](https://our.internmc.facebook.com/intern/diff/D19146669)

[ghstack-poisoned]
